File: backend/src/main/java/com/cotisapp/service/JournalUtilisateurService.java
package com.cotisapp.service;

import com.cotisapp.domain.entity.JournalAudit;
import com.cotisapp.domain.enums.TypeEvenementJournal;
import com.cotisapp.dto.request.EnregistrerEvenementJournalRequest;
import com.cotisapp.dto.response.JournalUtilisateurResponse;
import com.cotisapp.repository.JournalAuditRepository;
import jakarta.servlet.http.HttpServletRequest;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
@RequiredArgsConstructor
public class JournalUtilisateurService {

    private final JournalAuditRepository journalAuditRepository;
    private final JournalService journalService;

    @Transactional(readOnly = true)
    public Page<JournalUtilisateurResponse> lister(
            Long orgId,
            Long utilisateurId,
            TypeEvenementJournal type,
            Boolean succes,
            String search,
            int page,
            int size) {
        Pageable pageable = PageRequest.of(Math.max(0, page), Math.min(100, Math.max(1, size)));
        return journalAuditRepository
                .rechercherPourOrganisation(orgId, utilisateurId, type, succes, normaliserSearch(search), pageable)
                .map(this::toResponse);
    }

    @Transactional(readOnly = true)
    public long compter(Long orgId) {
        return journalAuditRepository.countByOrganisationId(orgId);
    }

    @Transactional
    public void enregistrerEvenementClient(
            Long orgId, EnregistrerEvenementJournalRequest request, HttpServletRequest httpRequest) {
        if (request.getTypeEvenement() == TypeEvenementJournal.MODULE_VISITE
                || request.getTypeEvenement() == TypeEvenementJournal.NAVIGATION) {
            journalService.enregistrerVisiteModule(
                    orgId,
                    request.getModuleCode(),
                    request.getModuleLibelle(),
                    request.getRoutePath(),
                    request.getDetails(),
                    httpRequest);
            return;
        }
        String action = request.getAction();
        journalService.enregistrer(JournalAudit.builder()
                .organisationId(orgId)
                .action(action)
                .typeEvenement(request.getTypeEvenement())
                .moduleCode(request.getModuleCode())
                .moduleLibelle(request.getModuleLibelle())
                .routePath(request.getRoutePath())
                .details(JournalAuditLibelleFormatter.enrichirDetailsAction(action, request.getDetails()))
                .ipAddress(JournalService.extraireIp(httpRequest))
                .userAgent(JournalService.tronquer(
                        httpRequest != null ? httpRequest.getHeader("User-Agent") : null, 500))
                .succes(request.getSucces() == null || request.getSucces()));
    }

    private JournalUtilisateurResponse toResponse(JournalAudit j) {
        return JournalUtilisateurResponse.builder()
                .id(j.getId())
                .organisationId(j.getOrganisationId())
                .utilisateurId(j.getUtilisateurId())
                .utilisateurEmail(j.getUtilisateurEmail())
                .utilisateurNom(j.getUtilisateurNom())
                .role(j.getRole())
                .membreId(j.getMembreId())
                .action(j.getAction())
                .typeEvenement(j.getTypeEvenement())
                .typeEvenementLibelle(JournalAuditLibelleFormatter.libelleTypeCourt(j.getTypeEvenement()))
                .moduleCode(j.getModuleCode())
                .moduleLibelle(j.getModuleLibelle())
                .routePath(j.getRoutePath())
                .details(JournalAuditLibelleFormatter.detailAffichage(j))
                .ipAddress(j.getIpAddress())
                .userAgent(j.getUserAgent())
                .navigateurResume(JournalAuditLibelleFormatter.resumeNavigateur(j.getUserAgent()))
                .succes(j.getSucces())
                .dateCreation(j.getDateCreation())
                .libelleResume(JournalAuditLibelleFormatter.libelleResume(j))
                .build();
    }

    private static String normaliserSearch(String search) {
        if (search == null) {
            return null;
        }
        String s = search.trim();
        return s.isEmpty() ? null : s;
    }
}

File: frontend/src/app/core/services/journal-utilisateur.service.ts
import { Injectable, inject } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { environment } from '../../../environments/environment';

export type TypeEvenementJournal =
  | 'CONNEXION'
  | 'DECONNEXION'
  | 'CONNEXION_ECHEC'
  | 'MODULE_VISITE'
  | 'ACTION_METIER'
  | 'NAVIGATION'
  | 'SECURITE';

export interface JournalUtilisateurDto {
  id: number;
  organisationId: number | null;
  utilisateurId: number | null;
  utilisateurEmail: string | null;
  utilisateurNom: string | null;
  role: string | null;
  membreId: number | null;
  action: string;
  typeEvenement: TypeEvenementJournal;
  typeEvenementLibelle: string;
  moduleCode: string | null;
  moduleLibelle: string | null;
  routePath: string | null;
  details: string | null;
  ipAddress: string | null;
  userAgent: string | null;
  navigateurResume: string | null;
  succes: boolean;
  dateCreation: string;
  libelleResume: string;
}

export interface JournalPageDto {
  content: JournalUtilisateurDto[];
  totalElements: number;
  totalPages: number;
  number: number;
  size: number;
}

export interface EnregistrerEvenementJournalBody {
  typeEvenement: TypeEvenementJournal;
  action: string;
  moduleCode?: string;
  moduleLibelle?: string;
  routePath?: string;
  details?: string;
  succes?: boolean;
}

export const TYPES_EVENEMENT_JOURNAL: { value: TypeEvenementJournal | ''; label: string }[] = [
  { value: '', label: 'Tous les types' },
  { value: 'CONNEXION', label: 'Connexion' },
  { value: 'DECONNEXION', label: 'Déconnexion' },
  { value: 'CONNEXION_ECHEC', label: 'Échec connexion' },
  { value: 'MODULE_VISITE', label: 'Module visité' },
  { value: 'ACTION_METIER', label: 'Action métier' },
  { value: 'NAVIGATION', label: 'Navigation' },
  { value: 'SECURITE', label: 'Sécurité' },
];

@Injectable({ providedIn: 'root' })
export class JournalUtilisateurService {
  private readonly http = inject(HttpClient);

  lister(
    orgId: number,
    options?: {
      utilisateurId?: number;
      type?: TypeEvenementJournal;
      succes?: boolean;
      search?: string;
      page?: number;
      size?: number;
    }
  ) {
    let params = new HttpParams()
      .set('page', String(options?.page ?? 0))
      .set('size', String(options?.size ?? 30));
    if (options?.utilisateurId != null) {
      params = params.set('utilisateurId', String(options.utilisateurId));
    }
    if (options?.type) {
      params = params.set('type', options.type);
    }
    if (options?.succes != null) {
      params = params.set('succes', String(options.succes));
    }
    if (options?.search?.trim()) {
      params = params.set('search', options.search.trim());
    }
    return this.http.get<JournalPageDto>(
      `${environment.apiUrl}/organisations/${orgId}/journal-utilisateur`,
      { params }
    );
  }

  compter(orgId: number) {
    return this.http.get<{ total: number }>(
      `${environment.apiUrl}/organisations/${orgId}/journal-utilisateur/count`
    );
  }

  enregistrerEvenement(orgId: number, body: EnregistrerEvenementJournalBody) {
    return this.http.post<void>(
      `${environment.apiUrl}/organisations/${orgId}/journal-utilisateur/evenement`,
      body
    );
  }
}
